rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,363 @@
+
+## `callPackage` with `nix-build -E`
+
+A commonly recommended way of building package directories in Nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
+Since the path changes `package.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/package.nix {}'`, which is harder for users.
+However, calling a path like this is an anti-pattern anyway, because it doesn't use the correct Nixpkgs version and it doesn't use the correct argument overrides.
+The correct way of doing it was to add the package to `all-packages.nix`, then calling `nix-build -A hello`.
+This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
+By teaching users that `pkgs/unit/<shard>/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.
+
+## Manual removal of custom arguments
+
+While this RFC allows passing custom arguments, doing so means that `all-packages.nix` will have to be maintained for that package.
+In specific cases where attributes of custom arguments are of the form `name = value` and `name` isn't a package attribute, they can be avoided without breaking the API.
+To do so, ensure that the function in the called file has `value` as an argument and set the default of the `name` argument to `value`.
+
+This notably doesn't work when `name` is already a package attribute, because then the default is never used and instead overridden.
+
+## Package variants
+
+Sometimes there's a need to create a variant of a package with different `callPackage` arguments. This can be achieved using `.override` as follows:
+```nix
+{
+  graphviz_nox = graphviz.override { withXorg = false; };
+}
+```
+
+However this can cause problems with an overlay that tries to make the variant the default as follows:
+```nix
+self: super: {
+  # Oops, infinite recursion!
+  graphviz = self.graphviz_nox;
+}
+```
+
+Because of this, there's the pattern of duplicating the `callPackage` call with the custom arguments as such:
+```nix
+{
+  graphviz_nox = callPackage ../tools/graphics/graphviz { withXorg = false; };
+}
+```
+
+The semantics of how unit directories are checked by CI do allow the definition of package variants from unit directories:
+```nix
+{
+  graphviz_nox = callPackage ../unit/gr/graphviz/package.nix { withXorg = false; };
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This standard can only be used for top-level packages using `callPackage`, so not for e.g. `python3Packages.requests` or a package defined using `haskellPackages.callPackage`
+- It's not possible anymore to be a [GitHub code owner](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) of category directories.
```

Comment:
I think that's not a big problem, given that we already have a hacky way to notify people with the `meta.maintainers`.

Reply:
The CODEOWNERS feature is superior to our `meta.maintainers` in a couple of ways, especially since the latter is not always reliable in all circumstances unfortunately. But as much as I am [personally a fan](https://github.com/NixOS/nixpkgs/issues/143441) of using the CODEOWNERS feature more (or an equivalent implemented in GitHub actions), I still think the benefits of this RFC outweigh the cost here.

Reply:
The CODEOWNERS feature is superior to our `meta.maintainers` in a couple of ways, especially since the latter is not always reliable in all circumstances unfortunately. But as much as I am [personally a fan](https://github.com/NixOS/nixpkgs/issues/143441) of using the CODEOWNERS feature more (or an equivalent implemented in GitHub actions), I still think the benefits of this RFC outweigh the cost here.